plan/physical.go

```diff
@@ -69,46 +67,40 @@ func (pp *physicalPlanner) Plan(spec *PlanSpec) (*PlanSpec, error) {
 	return final, nil
 }
 
-// TODO: This procedure should be encapsulated in a yield rewrite rule
-func removeYields(plan *PlanSpec) (*PlanSpec, error) {
-	for root := range plan.Roots {
-
-		name := DefaultYieldName
-
-		if yield, ok := root.ProcedureSpec().(YieldProcedureSpec); ok {
-
-			name = yield.YieldName()
-
-			if len(root.Predecessors()) != 1 {
-				return nil, errors.New("yield must have exactly one predecessor")
-			}
-
-			if _, ok := plan.Results[name]; ok {
-				return nil, fmt.Errorf("found duplicate yield name %q", name)
-			}
+type removeYieldRule struct {
+}
 
-			newRoot := root.Predecessors()[0]
-			newSucc := make([]PlanNode, 0, len(newRoot.Successors()))
+func (removeYieldRule) Name() string {
+	return "removeYield"
+}
 
-			for _, succ := range newRoot.Successors() {
-				if succ != root {
-					newSucc = append(newSucc, succ)
-				}
-			}
+func (removeYieldRule) Pattern() Pattern {
+	return YieldPat(Any())
+}
 
-			newRoot.ClearSuccessors()
-			newRoot.AddSuccessors(newSucc...)
-			plan.Replace(root, newRoot)
-			plan.Results[name] = newRoot
-			continue
-		}
+func (removeYieldRule) Rewrite(pn PlanNode) (PlanNode, bool) {
```

Comment:
This is just a comment. Rewrite rules are technically supposed to represent equivalence classes, and so the order in which rules are applied should not matter. However for yield removal it does - it must be the last rewrite step.

I'm slightly conflicted. On the one hand I really like how concise it is to express the process of removing yields with a single rule. But at the same time, I'm not sure if I like it being a rewrite rule due to it not preserving query behavior.

The rule itself is internal though. If it were a public facing rule I might have reservations, but because it's not, I think I'm fine with it being written this way.

Reply:
Personally I think if it's only used internally, it's okay.  But I'm glad that you're considering stuff like this.

Reply:
Yeah probably a bit over-thinking on my part.